### ISSUE TITLE
How to skip NA values when using summary_table with qwraps2 package?

### ISSUE THREAD

**Issue by WeCza** [dewittpe/qwraps2#106]
https://stackoverflow.com/q/70758071/1104685

I am trying to make a Baseline Characteristics Tables with qwraps2. My data is :

```
> str(joined_df2)
'data.frame':   259 obs. of  23 variables:
 $ SUBJID  : chr  "S001011" "S001013" "S001016" "S001017" ...
 $ AGE     : num  72 74 65 46 59 71 71 64 63 58 ...
 $ AGEU    : chr  "YEARS" "YEARS" "YEARS" "YEARS" ...
 $ FASFL.x : chr  "Y" "Y" "Y" "Y" ...
 $ SAFFL   : chr  "Y" "Y" "Y" "Y" ...
 $ TRT01P  : chr  "Treatment B" "Treatment A" "Treatment B" "Treatment B" ...
 $ HGTBL   : num  1.68 1.57 1.73 1.8 1.78 ...
 $ HGTBLU  : chr  "m" "m" "m" "m" ...
 $ WGTBL   : num  224 187 70.7 123.9 70.9 ...
 $ WGTBLU  : chr  "lb" "lb" "kg" "kg" ...
 $ DIABDUR : num  8 22 20 6 9 7 12 12 6 5 ...
 $ DIABDURU: chr  "years" "years" "years" "years" ...
 $ FASFL.y : chr  "Y" "Y" "Y" "Y" ...
 $ TRTP    : chr  "Treatment B" "Treatment A" "Treatment B" "Treatment B" ...
 $ AVISIT  : chr  "Visit 10 (Week 0)" "Visit 10 (Week 0)" "Visit 10 (Week 0)" "Visit 10 (Week 0)" ...
 $ VISITNUM: num  10 10 10 10 10 10 10 10 10 10 ...
 $ PARAM   : chr  "HbA1c Blood (%)" "HbA1c Blood (%)" "HbA1c Blood (%)" "HbA1c Blood (%)" ...
 $ PARAMCD : chr  "C64849B" "C64849B" "C64849B" "C64849B" ...
 $ AVAL    : num  8.6 8.4 7 7.3 8.2 7.7 7.3 8.8 7.3 8.4 ...
 $ AVALU   : chr  "%" "%" "%" "%" ...
 $ ANL01FL : chr  "Y" "Y" "Y" "Y" ...
 $ ANL01REA: chr  NA NA NA NA ...
 $ TRTP2   : chr  "Treatment B" "Treatment A" "Treatment B" "Treatment B" ...

I would like to include mean (SD), median, min, max value of the variable AGE and group it by TRTP2 . The variable AGE contains two NA values:
When I am running a code, I am getting an error:

> library(qwraps2)
> options(qwraps2_markup = 'markdown') # default is latex

> joined_df2_summaries <-
+   list("Age (yrs)" =
+          list(
+            "Mean (SD)"    = ~ qwraps2::mean_sd(AGE, denote_sd = "paren"),
+            "Median"       = ~ qwraps2::median_iqr(AGE),
+            "Min:"         = ~ min(AGE),
+            "Max:"         = ~ max(AGE)))
> summary_table(dplyr::group_by(joined_df2, TRTP2), joined_df2_summaries)

Error in quantile.default(x, probs = c(1, 3)/4, na.rm = na_rm) : 
  missing values and NaN's not allowed if 'na.rm' is FALSE

I tried to use na.rm=TRUE inside, but it does not help:

> joined_df2_summaries <-
+   list("Age (yrs)" =
+          list(
+            "Mean (SD)"    = ~ qwraps2::mean_sd(AGE, denote_sd = "paren", na.rm=TRUE),
+            "Median"       = ~ qwraps2::median_iqr(AGE, na.rm=TRUE),
+            "Min:"         = ~ min(AGE, na.rm=TRUE),
+            "Max:"         = ~ max(AGE, na.rm=TRUE)))
> summary_table(dplyr::group_by(joined_df2, TRTP2), joined_df2_summaries)
Error in qwraps2::mean_sd(AGE, denote_sd = "paren", na.rm = TRUE) : 
  unused argument (na.rm = TRUE)
```

How can I calculate mean, etc. of AGE excluding NA values? 

---

By submitting and issue I affirm:

- [ ] Bug Report
    - [ ] I have included a minimal, self-contained, and reproducible example.  (Use of the R package [reprex](https://cran.r-project.org/package=reprex) is recommended for creating reproducible examples.)
    - [ ] Included session info via base R `sessionInfo()`, `sessioninfo::session_info()`, or `xfun::session_info()`.
    - [ ] I have updated all packages.
    - [ ] I have tried the development version: `remotes::install_github("dewittpe/qwraps2")`

- [ ] Question
    - [x] I have cross posted the question on Stack Overflow or RStudio Community
      and included a link to that posting within the post here.

- [ ] Feature Request
    - [ ] I have described the use case for the feature and explained why it
      would be useful.
    - [ ] I have provided a reprex sufficient to outline and inform the feature.

- [ ] I have formatted the issue via the Github Markdown syntax for to make the
  issue easy to read and examples easy to evaluate.

- [ ] I understand that my issue may be closed if I fail to provide the
  requested details.

**Comment by dewittpe** [comment]
There are multiple ways to deal with missing values.  I strongly encourage you to consider why there are missing values and if they should be reported or not.

Functions such as `qwraps2::mean_sd` or `qwraps2::n_perc`will report values including a summary of known values.

You could wrap the variable name in `na.omit`, use `na_rm = TRUE`.

**Comment by dewittpe** [comment]
@WeCza the argument for ignoring missing values in `qwraps2::mean_sd` and `qwraps2::median_iqr` is _not_ `na.rm` it _is_ `na_rm`.

**Comment by WeCza** [comment]
@dewittpe Thank you so much for your quick reply. It worked. However, I encountered another problem. When I run summary_table, the output contains lots of nbsp, which I believe is something called non-breaking space. Do you have any suggestion, why I get this character ?
![NBS](https://user-images.githubusercontent.com/97986816/150220905-448d39fe-fdf4-4ffb-be7b-364efa9e7114.jpg)

**Comment by WeCza** [comment]
Is it possible that if I save this table to e.g., PDF file, there will be no nbsp?

**Comment by dewittpe** [comment]
The non-breaking space is there for formatting reasons.  There might be better ways to do this but it works for me and that's why it is part of the printing method for `summary_table`.   Once the markdown is rendered as html, or if tex is compiled via LaTeX, then the output looks nice:



  | Treatment A (N = 129) | Treatment B (N = 130)
-- | -- | --
**Age (yrs)** |   |  
&nbsp;&nbsp; Mean (SD) | 59.98 (8.50) | 59.29 (9.66)
&nbsp;&nbsp; Median | 61.00 (55.00, 65.00) | 60.00 (54.00, 66.00)
&nbsp;&nbsp; Min: | 36 | 29
&nbsp;&nbsp; Min: | 36 | 29

**Comment by WeCza** [comment]
Right! Thank you. If I may, could I ask one more question? Is there a way to add a footnote right below the table in the printing method? I have been trying footnote, add_footnote but they are not found as functions.

**Comment by dewittpe** [comment]
> Is there a way to add a footnote right below the table in the printing method? I have been trying footnote, add_footnote but they are not found as functions.


Not directly in `qwraps2`.  There are some methods in other packages that will add footnotes.  Honestly, when I need to added footnotes I do so explicitly in the markdown and LaTeX around a R chunk.

**Comment by WeCza** [comment]
Good suggestion.  Thank you! 

Everything has been working fine for me, but I think I did some manipulation to the code, while I wanted to try something else and now I can not render my table properly to html. I get this message:
[WARNING] This document format requires a nonempty <title> element.
  Defaulting to 'table.knit' as the title.
  To specify a title, use 'title' in metadata or --metadata title="...".

Output created: table.html


Could it be a problem with the summary_table or print method?

**Comment by dewittpe** [comment]
> [WARNING] This document format requires a nonempty <title> element.
Defaulting to 'table.knit' as the title.
To specify a title, use 'title' in metadata or --metadata title="...".
>
>Output created: table.html
>
>Could it be a problem with the summary_table or print method?

This looks like output form pandoc.  I'd need a reproducible example to troubleshoot this cause of the warning.

**Comment by WeCza** [comment]
> library(readxl)
> library(qwraps2)
> trail <- read_excel("C:/qwraps_test/trail.xlsx")
> trail
# A tibble: 10 x 3
   SUBJID      AGE TRT01P     
   <chr>     <dbl> <chr>      
 1 S00101010    72 Treatment B
 2 S00101313    74 Treatment A
 3 S00101616    65 Treatment B
 4 S00101717    66 Treatment B
 5 S00101818    59 Treatment A
 6 S00102121    88 Treatment A
 7 S00102222    97 Treatment B
 8 S00102828    54 Treatment A
 9 S00103131    63 Treatment B
10 S00103232    58 Treatment A
> trail.df<- as.data.frame(trail)
> trail.df
      SUBJID AGE      TRT01P
1  S00101010  72 Treatment B
2  S00101313  74 Treatment A
3  S00101616  65 Treatment B
4  S00101717  66 Treatment B
5  S00101818  59 Treatment A
6  S00102121  88 Treatment A
7  S00102222  97 Treatment B
8  S00102828  54 Treatment A
9  S00103131  63 Treatment B
10 S00103232  58 Treatment A
> options(qwraps2_markup = 'markdown')
> trail.df_summaries <-
+   list("Age (yrs)" =list
+        ("Mean (SD)"= ~ qwraps2::mean_sd(na.omit(AGE, na_rm=TRUE), denote_sd = "paren"),
+          "Median"= ~ qwraps2::median_iqr(na.omit(AGE, na_rm=TRUE)),
+          "Min:"= ~ min((na.omit(AGE, na_rm=TRUE))),
+          "Min:"= ~ min((na.omit(AGE, na_rm=TRUE)))))

> summary_table(dplyr::group_by(trail.df, TRT01P), trail.df_summaries)

> a <-cbind(summary_table(dplyr::group_by(trail.df, TRT01P), trail.df_summaries),
+       summary_table(trail.df, trail.df_summaries))
> print(a,
+ caption = "Table 1 - Full analysis set",
+ cnames = c("Treat A", "Treat B", "Total"))


The most strange thing is that, when I insert this code to R markdown and want to render it to html, the table does not render. 


![error](https://user-images.githubusercontent.com/97986816/150598713-bc6adcad-d3b7-404e-ba1f-108d179c361e.png)

**Comment by WeCza** [comment]
So this is my data: 

> trail.df
      SUBJID AGE      TRT01P
1  S00101010  72 Treatment B
2  S00101313  74 Treatment A
3  S00101616  65 Treatment B
4  S00101717  66 Treatment B
5  S00101818  59 Treatment A
6  S00102121  88 Treatment A
7  S00102222  97 Treatment B
8  S00102828  54 Treatment A
9  S00103131  63 Treatment B
10 S00103232  58 Treatment A

And the code
`library(readxl)
library(qwraps2)

trail <- read_excel("C:/qwraps_test/trail.xlsx")
trail

trail.df<- as.data.frame(trail)
trail.df

options(qwraps2_markup = 'markdown')

trail.df_summaries <-
  list("Age (yrs)" =list
       ("Mean (SD)"= ~ qwraps2::mean_sd(na.omit(AGE, na_rm=TRUE), denote_sd = "paren"),
         "Median"= ~ qwraps2::median_iqr(na.omit(AGE, na_rm=TRUE)),
         "Min:"= ~ min((na.omit(AGE, na_rm=TRUE))),
         "Min:"= ~ min((na.omit(AGE, na_rm=TRUE)))))


summary_table(dplyr::group_by(trail.df, TRT01P), trail.df_summaries)


a <-cbind(summary_table(dplyr::group_by(trail.df, TRT01P), trail.df_summaries),
      summary_table(trail.df, trail.df_summaries))
      
      
print(a,
caption = "Table 1 - full analysis set",
cnames = c("Treat A", "Treat B", "Total"))`


The most strange thing is that, when I insert this code to R markdown and want to render it to html, the table does not render.
![error](https://user-images.githubusercontent.com/97986816/150599852-4faee759-1dc7-4731-98e2-c5484ad3924b.png)

**Comment by dewittpe** [comment]
Are you working in an .Rmd file?   I added an [example Rmd file](https://github.com/dewittpe/qwraps2/blob/9e96d4f4ad658c01cebc9c56541dd8dd515b9490/examples/github-issues/qwraps2-issue-0106.Rmd) to the repo and it appears to compile as expected:  
![image](https://user-images.githubusercontent.com/3136617/150608680-d0ae1003-c2e6-4ba0-ac6d-c5611ac64d55.png)

This was built with pandoc version 2.17.0.1

and session info:
```r
R version 4.1.2 (2021-11-01)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.6.2

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] colorout_1.2-2

loaded via a namespace (and not attached):
 [1] compiler_4.1.2  fastmap_1.1.0   htmltools_0.5.2 tools_4.1.2     yaml_2.2.1      rmarkdown_2.11  knitr_1.37     
 [8] xfun_0.29       digest_0.6.29   rlang_0.4.12    evaluate_0.14  
```

**Comment by WeCza** [comment]
This was more than helpful! Thank you! Now I know everything. 

The warning about the title could be due to the pandoc version. My pandoc is  ‘2.14.0.3’, I will try to update it later and see whether the warning comes back. 

But the most important is that my table was not rendering to html because I only used {r}, and not {r, results="asis"}. Now, when I changed it, everythings renders correctly. That happens, when someone is "new" to something.

Thanks again for all your help! qwraps2 is a cool package for creating tables!
